File: 01/two.hs
import qualified Data.Set as Set

-- | Parse a frequency shift such as "+12" or "-15". 'read' chokes on a
-- leading plus sign, so strip it off first.
parseShift :: String -> Int
parseShift ('+' : digits) = read digits
parseShift shift = read shift

parseShifts :: String -> [Int]
parseShifts = map parseShift . filter (not . null) . lines

-- | Find the first frequency that is reached twice. The shifts are cycled
-- endlessly and 'scanl' gives us the lazy list of running totals, so we just
-- walk it until we hit a total we have already seen.
calc :: [Int] -> Int
calc shifts = go (scanl (+) 0 (cycle shifts)) Set.empty
  where
    go (total : remaining) seen
      | total `Set.member` seen = total
      | otherwise = go remaining (Set.insert total seen)
    go [] _ = error "unreachable: the list of shifts is cycled endlessly"

main :: IO ()
main = do
  input <- readFile "input.txt"
  putStrLn ("FOUND: " ++ show (calc (parseShifts input)))

File: 01/two.rs
use std::collections::HashSet;
use std::fs;

fn calc(input: &str) -> i64 {
    let shifts: Vec<i64> = input
        .lines()
        .filter(|line| !line.trim().is_empty())
        .map(|line| line.trim().parse::<i64>().unwrap())
        .collect();

    let mut seen = HashSet::new();
    let mut total = 0;
    seen.insert(total);

    // Cycle through the shifts until a frequency is reached twice.
    for shift in shifts.iter().cycle() {
        total += shift;
        if !seen.insert(total) {
            return total;
        }
    }

    unreachable!("input list was empty")
}

fn main() {
    let input = fs::read_to_string("input.txt").expect("failed to read input.txt");
    println!("{}", calc(&input));
}
